core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1493,6 +1494,23 @@ class ReplicaManager(val config: KafkaConfig,
             error(s"Error while making broker the follower for partition $partition with leader " +
               s"$newLeaderBrokerId in dir $dirOpt", e)
             responseMap.put(partition.topicPartition, Errors.KAFKA_STORAGE_ERROR)
+          case e: ZooKeeperClientException =>
+            // Finish operations for leaderEpoch-updated partitions up to this point.
+            replicaFetcherManager.removeFetcherForPartitions(partitionsToMakeFollower.map(_.topicPartition))
+            partitionsToMakeFollower.foreach { partition =>
+              completeDelayedFetchOrProduceRequests(partition.topicPartition)
+            }
+            val makeFollowerPartitionsAndLeaderOffset = partitionsToMakeFollower.map { partition =>
+              val leader = metadataCache.getAliveBrokers.find(_.id == partition.leaderReplicaIdOpt.get).get
+                .brokerEndPoint(config.interBrokerListenerName)
+              val fetchOffset = partition.localLogOrException.highWatermark
+              partition.topicPartition -> InitialFetchState(leader, partition.getLeaderEpoch, fetchOffset)
+            }.toMap
+            replicaFetcherManager.addFetcherForPartitions(makeFollowerPartitionsAndLeaderOffset)
+            stateChangeLogger.info(s"Because a ZooKeeper client exception occurred, completed become follower " +
+              s"state change with correlation identifier $correlationId from epoch ${partition.getLeaderEpoch} only for " +
+              s"those leaderEpoch-updated partitions with leader $newLeaderBrokerId before ZooKeeper disconnect occurred.", e)
+            error(s"ZooKeeper client occurred while rendering a $partition's follower through $zkClient.'", e)
```

Comment:
After the error log, you will continue executing the logic from line 1517-1570, which will re-do the work you have done here. I think you can just get rid of line 1498-1509 completely and let the code run through the remaining logic.